File: fixtures/typescript/alpha.ts
// alpha.ts — source side of the TS Type-1 and Type-2 duplicate pairs.

import { Row, Total } from "./types";
import * as util from "./util";

// Type-1 duplicate: identical to `type1Identical` in beta.ts.
function type1Identical(rows: Row[]): Total {
    let sum = 0;
    let count = 0;
    for (const r of rows) {
        sum += r.value;
        count += 1;
    }
    const mean = count > 0 ? sum / count : 0;
    return { sum, count, mean };
}

// Type-2 duplicate: same shape as `type2Renamed` in beta.ts, differs
// only in local binding and parameter names.
const type2Original = (items: Row[]): Total => {
    let tot = 0;
    let cnt = 0;
    for (const item of items) {
        tot += item.value;
        cnt += 1;
    }
    const avg = cnt > 0 ? tot / cnt : 0;
    return { sum: tot, count: cnt, mean: avg };
};

// Unique arrow — must NOT bucket with the duplicates.
const uniqueAlpha = (): number => {
    const base = 7;
    return base * 2 + 1;
};

export class Aggregator {
    constructor(private readonly factor: number) {}

    // Method is itself a syntactic unit; body uses `this` which stays.
    scale(values: number[]): number {
        let acc = 0;
        for (const v of values) {
            acc += v * this.factor;
        }
        return acc;
    }
}

export { type1Identical, type2Original, uniqueAlpha };

File: fixtures/python/alpha.py
# alpha.py -- source side of the Python Type-1 and Type-2 duplicate pairs.

from typing import List
from .types import Row, Total
import util.helpers


# Type-1 duplicate: identical to `type1_identical` in beta.py.
def type1_identical(rows: List) -> Total:
    sum_ = 0
    count = 0
    for r in rows:
        sum_ += r.value
        count += 1
    mean = sum_ / count if count > 0 else 0
    return Total(sum=sum_, count=count, mean=mean)


# Type-2 duplicate: same structure as `type2_shared` in beta.py,
# differs only in local binding and parameter names. Function name
# stays the same because Python has no anonymous function construct
# for this pattern and names are Kept by the profile.
def type2_shared(items: List) -> Total:
    tot = 0
    cnt = 0
    for item in items:
        tot += item.value
        cnt += 1
    avg = tot / cnt if cnt > 0 else 0
    return Total(sum=tot, count=cnt, mean=avg)


# Unique function -- must NOT bucket with the duplicates.
def unique_alpha() -> int:
    base = 7
    return base * 2 + 1


class Aggregator:
    def __init__(self, factor: int):
        self.factor = factor

    def scale(self, values: List) -> int:
        acc = 0
        for v in values:
            acc += v * self.factor
        return acc
